pkg/services/ngalert/state/historian/loki.go

```diff
@@ -30,11 +36,114 @@ func (h *RemoteLokiBackend) TestConnection() error {
 	return h.client.ping()
 }
 
-func (h *RemoteLokiBackend) RecordStatesAsync(ctx context.Context, _ *models.AlertRule, _ []state.StateTransition) {
+func (h *RemoteLokiBackend) RecordStatesAsync(ctx context.Context, rule *models.AlertRule, states []state.StateTransition) {
 	logger := h.log.FromContext(ctx)
-	logger.Debug("Remote Loki state history backend was called with states")
+	streams := h.buildStreams(rule, states, logger)
+	h.recordStreamsAsync(ctx, streams, logger)
 }
 
 func (h *RemoteLokiBackend) QueryStates(ctx context.Context, query models.HistoryQuery) (*data.Frame, error) {
 	return data.NewFrame("states"), nil
 }
+
+func (h *RemoteLokiBackend) buildStreams(rule *models.AlertRule, states []state.StateTransition, logger log.Logger) []stream {
+	buckets := make(map[string][]row) // label repr -> entries
+	for _, state := range states {
+		if !shouldRecord(state) {
+			continue
+		}
+
+		labels := removePrivateLabels(state.State.Labels)
+		labels["orgID"] = fmt.Sprint(rule.OrgID)
+		labels["ruleUID"] = fmt.Sprint(rule.UID)
+		labels["group"] = fmt.Sprint(rule.RuleGroup)
+		labels["folderUID"] = fmt.Sprint(rule.NamespaceUID)
+		repr := labels.String()
+
+		entry, err := lokiRepresentation(rule, state)
+		if err != nil {
+			logger.Error("Failed to construct history record for state, skipping", "error", err)
+			continue
+		}
+		buckets[repr] = append(buckets[repr], row{
+			At:  state.State.LastEvaluationTime,
+			Val: entry,
+		})
+	}
+
+	result := make([]stream, 0, len(buckets))
+	for repr, rows := range buckets {
+		labels, err := data.LabelsFromString(repr)
+		if err != nil {
+			logger.Error("Failed to parse frame labels, skipping state history batch: %w", err)
+			continue
+		}
+		result = append(result, stream{
+			Stream: labels,
+			Values: rows,
+		})
+	}
+
+	return result
+}
+
+func (h *RemoteLokiBackend) recordStreamsAsync(ctx context.Context, streams []stream, logger log.Logger) {
+	go func() {
+		if err := h.recordStreams(ctx, streams, logger); err != nil {
+			logger.Error("Failed to save alert state history batch", "error", err)
+		}
+	}()
+}
+
+func (h *RemoteLokiBackend) recordStreams(ctx context.Context, streams []stream, logger log.Logger) error {
+	logger.Error("pushing streams", "streams", streams)
```

Comment:
I don't think this should be `Error`
```suggestion
	logger.Debug("pushing streams", "streams", streams)
```

Reply:
I just removed it entirely. It logs the full structs, which is also bad.